### PR TITLE
use CVODES error string instead of numeric flag in error messages

### PR DESCRIPTION
**Pull request by yizhang-yiz**
This PR addresses #1039 by replacing flag with strings from CVODES.

## Summary

The replacement happens inside the checking functions. With added `cvodes_check` we call the library functions that map flag to error string constant.

## Tests

All CVODES  unit tests with error throw checks.

## Side Effects

na

## Release notes
More informative error messages for ODE solvers.

## Checklist

- [x] Math issue #1039.

- [x] Copyright holder:  Metrum Research Group

    By submitting this pull request, the copyright holder is agreeing to the license the submitted work under the following licenses:
      - Code: BSD 3-clause (https://opensource.org/licenses/BSD-3-Clause)
      - Documentation: CC-BY 4.0 (https://creativecommons.org/licenses/by/4.0/)

- [x] the basic tests are passing

    - unit tests pass (to run, use: `./runTests.py test/unit`)
    - header checks pass, (`make test-headers`)
    - dependencies checks pass, (`make test-math-dependencies`)
    - docs build, (`make doxygen`)
    - code passes the built in [C++ standards](https://github.com/stan-dev/stan/wiki/Code-Quality) checks (`make cpplint`)

- [x] the code is written in idiomatic C++ and changes are documented in the doxygen

- [x] the new changes are tested

### CODE REVIEW

**Comment by yizhang-yiz** [comment]
@syclik do you want to take a look of this one?

**Comment by yizhang-yiz** [comment]
@wds15 would you like to review it? thx

**Comment by yizhang-yiz** [comment]
This is ready for review. @SteveBronder @rok-cesnovar any of you have time for this? It's a straightforward impl.

File: stan/math/prim/err/check_flag_sundials.hpp [hunk 1]
```diff
@@ -1,12 +1,178 @@
 #ifndef STAN_MATH_PRIM_ERR_CHECK_FLAG_SUNDIALS_HPP
 #define STAN_MATH_PRIM_ERR_CHECK_FLAG_SUNDIALS_HPP
 
+#include <kinsol/kinsol.h>
+#include <cvodes/cvodes.h>
 #include <stan/math/prim/meta.hpp>
-#include <stan/math/prim/err/throw_domain_error.hpp>
+#include <stan/math/prim/err/domain_error.hpp>
```
Comment (rok-cesnovar) [inline stan/math/prim/err/check_flag_sundials.hpp:7]: Please reorder this to:

#include <stan/math/prim/meta.hpp>
#include <stan/math/prim/err/throw_domain_error.hpp>
#include <stan/math/prim/err/domain_error.hpp>
#include <kinsol/kinsol.h>
#include <cvodes/cvodes.h>


The order we generally try to follow is: Stan Math master header file(like meta.hpp), individual Stan Math header files, std:: header files, other C++ header files, other C header files.

This isn't always possible, so not a strict thing we automatically check,

File: stan/math/prim/err/check_flag_sundials.hpp [hunk 2]
```diff
@@ -15,34 +181,61 @@ namespace math {
...
+ * until it's fixed we cannot use it to extract flag error string.
+ *
+ * @param flag Error flag
+ * @param func_name calling function name
+ * @throw <code>std::runtime_error</code> if the flag is negative.
+ */
+inline void kinsol_check(int flag, const char* func_name) {
+  std::ostringstream ss;
+  if (flag < 0) {
+    ss << "algebra_solver failed with error flag " << flag << ".";
```
Comment (rok-cesnovar) [inline stan/math/prim/err/check_flag_sundials.hpp:211]:
```suggestion
  if (flag < 0) {
    std::ostringstream ss;
    ss << "algebra_solver failed with error flag " << flag << ".";
```

File: stan/math/prim/err/check_flag_sundials.hpp [hunk 3]
```diff
@@ -15,34 +181,61 @@ namespace math {
...
- *   should take before throwing an error
- * @throw <code>std::domain_error</code> if flag means maximum number of
- *   iterations exceeded in the algebra solver.
- * @throw <code>std::runtime_error</code> if the flag is negative for
- *   any other reason.
+ * @param func_name calling function name
+ * @param max_num_steps max number of nonlinear iters
+ * @throw <code>std::runtime_error</code> if the flag is negative.
+ * @throw <code>std::domain_error</code> if the flag indicates max
+ * number of steps is exceeded..
```
Comment (rok-cesnovar) [inline stan/math/prim/err/check_flag_sundials.hpp:227]:
```suggestion
 * number of steps is exceeded.
```

File: stan/math/rev/functor/cvodes_utils.hpp [hunk 4]
```diff
@@ -10,38 +10,34 @@
...
+// extern "C" inline void cvodes_err_handler(int error_code, const char* module,
+//                                           const char* function, char* msg,
+//                                           void* eh_data) {
+//   if (error_code != CV_TOO_MUCH_WORK) {
+//     std::ostringstream msg1;
+//     msg1 << msg << " Error code: ";
+
+//     throw_domain_error(module, function, error_code, msg1.str().c_str());
+//   }
+// }
```
Comment (rok-cesnovar) [inline stan/math/rev/functor/cvodes_utils.hpp:22]: If this code is not needed, lets just remove it.

File: stan/math/rev/functor/cvodes_utils.hpp [hunk 5]
```diff
@@ -10,38 +10,34 @@
...
 
 inline void cvodes_set_options(void* cvodes_mem,
                                // NOLINTNEXTLINE(runtime/int)
                                long int max_num_steps) {
   // forward CVode errors to noop error handler
-  CVodeSetErrHandlerFn(cvodes_mem, cvodes_err_handler, nullptr);
+  // CVodeSetErrHandlerFn(cvodes_mem, cvodes_err_handler, nullptr);
```
Comment (rok-cesnovar) [inline stan/math/rev/functor/cvodes_utils.hpp:28]: Same here.

File: stan/math/prim/err/check_flag_sundials.hpp [hunk 6]
```diff
@@ -1,12 +1,178 @@
...
+    case -11:
+      msg = {"CV_UNREC_RHSFUNC_ERR",
+             "The right-hand side function had a recoverable error, but no "
+             "recovery is possible"};
+      break;  // NOLINT
+    case -27:
+      msg = {"CV_TOO_CLOSE",
+             "The output and initial times are too close to each other"};
+      break;  // NOLINT
+    default:
```
Comment (rok-cesnovar) [inline stan/math/prim/err/check_flag_sundials.hpp:69]: I am not sure I completely understand why there is another switch inside the default? Shouldn't this just be a one-level switch-case statement?

File: stan/math/prim/err/check_flag_sundials.hpp [hunk 7]
```diff
@@ -15,34 +181,61 @@ namespace math {
...
+    }
+  }
+}
+
+/**
+ * Throws an exception message when the functions in KINSOL
+ * fails. When the exception is caused
+ * by a tuning parameter the user controls, gives a specific
+ * error. "KINGetReturnFlagName()" from sundials has a mem leak bug so
+ * until it's fixed we cannot use it to extract flag error string.
```
Comment (rok-cesnovar) [inline stan/math/prim/err/check_flag_sundials.hpp:202]: This Doxygen comment does not align with the actual implementation that just returns the error number. Sundials should be capitalized as well.

File: stan/math/prim/err/check_flag_sundials.hpp [hunk 8]
```diff
@@ -15,34 +181,61 @@ namespace math {
...
  *
  * @param flag Error flag
- * @param max_num_steps Maximum number of iterations the algebra solver
- *   should take before throwing an error
- * @throw <code>std::domain_error</code> if flag means maximum number of
- *   iterations exceeded in the algebra solver.
- * @throw <code>std::runtime_error</code> if the flag is negative for
- *   any other reason.
+ * @param func_name calling function name
+ * @param max_num_steps max number of nonlinear iters
```
Comment (rok-cesnovar) [inline stan/math/prim/err/check_flag_sundials.hpp:224]: Let's just use one of the terms steps/iterations.

**Review by rok-cesnovar — CHANGES_REQUESTED** [review]
Overall I think this is a great improvement of the current state. Thanks @yizhang-yiz!

I commented in a few spots that things could be slightly improved.

File: stan/math/prim/err/check_flag_sundials.hpp (hunk 6 above)
Comment (yizhang-yiz) [reply stan/math/prim/err/check_flag_sundials.hpp:69]: I put most common cases at the top. The rest are in a 2nd level of switch. IMO it helps to read, may also help compiler optimize but in our case it doesn't really matter.

Comment (rok-cesnovar) [reply stan/math/prim/err/check_flag_sundials.hpp:69]: Makes sense.

File: stan/math/prim/err/check_flag_sundials.hpp (hunk 2 above)
Comment (yizhang-yiz) [reply stan/math/prim/err/check_flag_sundials.hpp]: done, thanks.

File: stan/math/rev/functor/cvodes_utils.hpp (hunk 4 above)
Comment (yizhang-yiz) [reply stan/math/rev/functor/cvodes_utils.hpp]: done, thanks.

File: stan/math/prim/err/check_flag_sundials.hpp (hunk 7 above)
Comment (yizhang-yiz) [reply stan/math/prim/err/check_flag_sundials.hpp:200]: cleaned up, thanks.

File: stan/math/prim/err/check_flag_sundials.hpp (hunk 8 above)
Comment (yizhang-yiz) [reply stan/math/prim/err/check_flag_sundials.hpp]: changed to "iterations".

**Review by rok-cesnovar — APPROVED** [review]
Looks great. Will probably take a day or a few to get Jenkins tests to pass again. Merge whenever the green light is back.

**Comment by yizhang-yiz** [comment]
thanks a lot @rok-cesnovar

**Pull request merged**